### PR TITLE
Get string representation of javadsl Content and Media Type

### PR DESCRIPTION
**Pull request by mdedetrich**
Currently in the javadsl its not possible to get the string representation of the `Content-Type`/`Media-Type` because its only the scaladsl that implements the `value` function (via the `ValueRenderable` trait). This PR adds the `value` method to the core Java interfaces (the `render` method is anyways implemented in the subclasses hence why this PR is compiling without any additional changes)

As a workaround I currently have to do this (its in kotlin but should be understandable)

```kotlin
internal fun contentTypeToString(contentType: ContentType): String {
    return when (contentType) {
        is org.apache.pekko.http.scaladsl.model.ContentType ->
            return contentType.value()
        else -> throw RuntimeException("Unsupported content type: ${contentType::class.qualifiedName}")
    }
}
```

This works because it just so happens that every javdsl `ContentType` is also a scaladsl `ContentType`

### CODE REVIEW

File: http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala
Comment (pjfanning) [inline http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:67]: can you add `@since 1.1.1` to the 2 new public functions?

Comment (pjfanning) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:68]: if we do go the 1.2.0 route, we will need to delay this PR merge until we get to create the 1.1.x branch and the CI jobs related to it

Comment (mdedetrich) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:68]: Thats fine, its not urgent

Comment (pjfanning) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:68]: * ok - let's go the 1.2.0 route
* I'll try to set up the CI builds over the next few days

**Review by pjfanning — APPROVED** [review]
lgtm - the 1.1.x branch exists now so this is safe to merge but I would appreciate a delay on the merge until we get #631 merged

**Comment by mdedetrich** [comment]
> lgtm - the 1.1.x branch exists now so this is safe to merge but I would appreciate a delay on the merge until we get #631 merged

I would wait for 1.2.x, there is a workaround and since we are adding methods to an interface we should be bumping the minor version

**Comment by pjfanning** [comment]
> > lgtm - the 1.1.x branch exists now so this is safe to merge but I would appreciate a delay on the merge until we get #631 merged
> 
> I would wait for 1.2.x, there is a workaround and since we are adding methods to an interface we should be bumping the minor version

* we use 'main' branch for the development version and 1.2.0 is becoming the development version
* future 1.1.x patch releases will be built using the '1.1.x' branch

File: http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala
Comment (pjfanning) [inline http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: I think this should be getValue. More Java API like. Other existing methods seem to be getters.

**Review by pjfanning — CHANGES_REQUESTED** [review]

File: http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala
Comment (mdedetrich) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: Let me check because this isn't consistent, i.e. there is `getCharsetOption` but then there is also `mediaType` (which if it was consistent would be `getMediaType`)

Comment (pjfanning) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: reasonable point - it does seem like the existing code can't decide on getter style versus field name style.

File: http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala
Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:68]: Maybe just override the `toString` method?

File: http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala
Comment (He-Pin) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: In Spring-web, there is a `MimeTypeUtils#toString`, for me this method should be `getValue` or just override the `toString` method.

Comment (jxnu-liguobin) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: You're right, the type is always String, toString is more in line with Java.

Comment (pjfanning) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: toString would have the useful side effect that we would be ok to merge this to v1.1.x patch line

Comment (mdedetrich) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: I don't think putting `toString` on the interface is a good idea, its something that should concretely be implemented by its sub types.

Comment (mdedetrich) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: So fyi, changing it to `.getValue()` is more contrived than initially anticipated. Naively you would do something like

```scala
def getValue: String = value
```

However the `org.apache.pekko.http.javadsl.model.ContentType` has no visibility for the `value` method within the `ValueRenderable` trait (after all this is the exact issue we are dealing with). In order to make this work `org.apache.pekko.http.javadsl.model.ContentType` would have to extend `ValueRenderable` which would also expose the `value` method since its public i.e.
```
/**
 * INTERNAL API
 *
 * An entity that has a rendered value (like an HttpHeader)
 */
@InternalApi
private[http] trait ValueRenderable extends ToStringRenderable {
  def value: String = toString
}
```
which means the trait would both have `value` and `getValue` methods, far from ideal (note that even though the trait is marked as `private[http]`, this is only visible to the Scala compiler, Java has no concept of it and hence it only sees it as a standard interface with a public method).

I think this discussion is revealing of what appears to be an overlooked design misdecision, which is that we should be having 2 `ValueRenderable` traits, one for Java api and one for Scala api and only the respective implementations extending those traits.

I can go ahead and do this change but its obviously going to be more work. Thankfully `ValueRenderable` trait is marked as internal/private which means I only need to take care that a public `def value: String = toString` exists for the Scala api.

Wdyt?

Comment (pjfanning) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: If leaving this as `def value = ???` is easier, I am ok with that. It seems like the Java API is already inconsistent stylistically.

Comment (mdedetrich) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: > If leaving this as `def value = ???` is easier, I am ok with that. It seems like the Java API is already inconsistent stylistically.

I personally would like to avoid that, I think that leaving unimplemented methods is taking it a bridge too far. @raboof what are your thoughts on this? I don't have an issue with doing this properly, this PR is slated for 1.2.0 anyways.

Comment (raboof) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:70]: I think it would make sense to keep this as `def value: String`. I don't think we need the `= ???` as all implementations of this trait have an implementation for this value, right?

Indeed 'getValue' would be more Java-esque, but since we already have `mediaType` this seems fine.

File: http-core/src/main/mima-filters/1.2.x.backwards.excludes/content-media-type-value-string.backwards.excludes
Comment (raboof) [inline http-core/src/main/mima-filters/1.2.x.backwards.excludes/content-media-type-value-string.backwards.excludes:2]: This is only justified if there are no user types extending from these classes (or their subclasses), i.e. if these traits are marked `@DoNotInherit`. The traits are not marked `@DoNotInherit`, so naively you'd say this change is not allowed. However, I think you can make the argument that the fact that `@DoNotInherit` is missing from this trait is an oversight, and it was never intended to be user-extended. It would be good to add the `@DoNotInherit` now, then?

Comment (mdedetrich) [reply http-core/src/main/mima-filters/1.2.x.backwards.excludes/content-media-type-value-string.backwards.excludes:2]: Do you mean putting `@DoNotInherit` on `ContentType`/`MediaType` traits themselves?

**Comment by mdedetrich** [comment]
> CustomMediaTypesExampleTest

Done

File: http-core/src/main/mima-filters/1.2.x.backwards.excludes/content-media-type-value-string.backwards.excludes
Comment (mdedetrich) [reply http-core/src/main/mima-filters/1.2.x.backwards.excludes/content-media-type-value-string.backwards.excludes:2]: No thats fair, I thought you were talking about `ValueRenderable` but that wouldn't make sense its already marked as `@InternalApi`.

We can remove and also fix the Java namings (i.e. `value` should be `getValue`) as well as removing `@DoNotInherit` in 2.x.x

File: http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala
Comment (mdedetrich) [reply http-core/src/main/scala/org/apache/pekko/http/javadsl/model/ContentType.scala:73]: Okay fair, I will mark this as resolved (if someone still disagrees then unresolve conversation).

**Comment by mdedetrich** [comment]
@pjfanning Can you re-review this PR given @raboof 's comments?

**Review by raboof — APPROVED** [review]

**Comment by pjfanning** [comment]
> @pjfanning Can you re-review this PR given @raboof 's comments?

The changes look ok. We still need to block the merge until we have the 1.1.x branch and related CI builds.

**Comment by raboof** [comment]
> We still need to block the merge until we have the 1.1.x branch and related CI builds.

I think I missed why we'd want to delay this merge - this PR as it now stands seems like a regular backwards-compatible change (assuming nobody inherited from `ContentType` or `MediaType`, which seems like a reasonable assumption)?

**Comment by pjfanning** [comment]
> > We still need to block the merge until we have the 1.1.x branch and related CI builds.
> 
> I think I missed why we'd want to delay this merge - this PR as it now stands seems like a regular backwards-compatible change (assuming nobody inherited from `ContentType` or `MediaType`, which seems like a reasonable assumption)?

The PR currently has `@since 1.2.0`. If these change to `@since 1.1.1` then we don't need to start preparing for 1.2.0 snapshots.

**Comment by raboof** [comment]
> > > We still need to block the merge until we have the 1.1.x branch and related CI builds.
> > 
> > I think I missed why we'd want to delay this merge - this PR as it now stands seems like a regular backwards-compatible change (assuming nobody inherited from `ContentType` or `MediaType`, which seems like a reasonable assumption)?
> 
> The PR currently has `@since 1.2.0`. If these change to `@since 1.1.1` then we don't need to start preparing for 1.2.0 snapshots.

Aah, gotcha. This begs the question whether those `@since` annotations are really worth having. I can easily imagine situations where it isn't clear yet whether the next release will be a minor, patch or major one. Should/have we discuss(ed) this on the mailinglist?

**Comment by mdedetrich** [comment]
> I think I missed why we'd want to delay this merge - this PR as it now stands seems like a regular backwards-compatible change (assuming nobody inherited from `ContentType` or `MediaType`, which seems like a reasonable assumption)?

So actually, the reason for putting it in 1.2.0 is that while its not a breaking compatible change it is a forwards breaking change (hence the MiMa changes) which is allowed by semver in minor version bumps (its only backwards compatibility thats disallowed, this has to be done in a major patch update).

**Review by pjfanning — APPROVED** [review]
approved for main branch merge - now that main branch is repurposed for v1.2 dev

**Comment by mdedetrich** [comment]
Thanks, I will go ahead and merge it.

**Pull request merged**
